src/ffi.rs

```diff
@@ -4,686 +4,621 @@
-        existing_nullifiers_len: *mut u32,
-    ) -> u8;
-
-    /// Fetches the current anchor.
-    fn fetch_anchor(anchor: *mut [u8; BlsScalar::SIZE]) -> u8;
-
-    /// Fetches the current stake for a key.
-    ///
-    /// The value, eligibility, reward and counter should be written in
-    /// sequence, little endian, to the given buffer. If there is no value and
-    /// eligibility, the first 16 bytes should be zero.
-    fn fetch_stake(
-        pk: *const [u8; PublicKey::SIZE],
-        stake: *mut [u8; StakeInfo::SIZE],
-    ) -> u8;
-
-    /// Request the node to prove the given unproven transaction.
-    fn compute_proof_and_propagate(
-        utx: *const u8,
-        utx_len: u32,
-        tx: *mut u8,
-        tx_len: *mut u32,
-    ) -> u8;
-
-    /// Requests the node to prove STCT.
-    fn request_stct_proof(
-        inputs: *const [u8; STCT_INPUT_SIZE],
-        proof: *mut [u8; Proof::SIZE],
-    ) -> u8;
-
-    /// Request the node to prove WFCT.
-    fn request_wfct_proof(
-        inputs: *const [u8; WFCT_INPUT_SIZE],
-        proof: *mut [u8; Proof::SIZE],
-    ) -> u8;
+/// Allocates a buffer of `len` bytes on the WASM memory.
+#[no_mangle]
+pub fn malloc(len: i32) -> i32 {
+    let bytes = vec![0u8; len as usize];
+    let ptr = bytes.as_ptr();
+    mem::forget(bytes);
+    ptr as i32
 }
 
-macro_rules! unwrap_or_bail {
-    ($e: expr) => {
-        match $e {
-            Ok(v) => v,
-            Err(e) => {
-                return Error::<FfiStore, FfiStateClient, FfiProverClient>::from(e).into();
+/// Frees a previously allocated buffer on the WASM memory.
+#[no_mangle]
+pub fn free_mem(ptr: i32, len: i32) {
+    let ptr = ptr as *mut u8;
+    let len = len as usize;
+    unsafe {
+        Vec::from_raw_parts(ptr, len, len);
+    }
+}
+
+/// Computes the total balance of the given notes.
+///
+/// The arguments are expected to be rkyv serialized [BalanceArgs] with a
+/// pointer defined via [malloc]. It will consume the `args` allocated region
+/// and drop it.
+#[no_mangle]
+pub fn balance(args: i32, len: i32) -> i32 {
+    let args = args as *mut u8;
+    let len = len as usize;
+    let args = unsafe { Vec::from_raw_parts(args, len, len) };
+
+    let BalanceArgs { seed, notes } = match rkyv::from_bytes(&args) {
+        Ok(a) => a,
+        Err(_) => return BalanceResponse::fail(),
+    };
+
+    let notes: Vec<Note> = match rkyv::from_bytes(&notes) {
+        Ok(n) => utils::sanitize_notes(n),
+        Err(_) => return BalanceResponse::fail(),
+    };
+
+    let mut keys = unsafe { [mem::zeroed(); MAX_KEY + 1] };
+    let mut values = Vec::with_capacity(notes.len());
+    let mut keys_len = 0;
+    let mut sum = 0u64;
+
+    'outer: for note in notes {
+        // we iterate all the available keys until one can successfully decrypt
+        // the note. if all fails, returns false
+        for idx in 0..=MAX_KEY {
+            if keys_len == idx {
+                keys[idx] = key::derive_vk(&seed, idx as u64);
```

Comment:
Do we have a gurantee here that indexing like that won't panic? Because panicking in a non rust context will lead to UB unless we catch it with  [`panic::catch_unwind`](https://doc.rust-lang.org/std/panic/fn.catch_unwind.html)

Reply:
We have since we declare this array a couple of lines before with `MAX_KEY+1` len